lib/SILOptimizer/Transforms/SILSROA.cpp

```diff
@@ -260,12 +265,32 @@ void SROAMemoryUseAnalyzer::chopUpAlloca(std::vector<AllocStackInst *> &Worklist
 
   // Change any aggregate stores into extracts + field stores.
   for (auto *SI : Stores) {
-    SILBuilderWithScope B(SI);
-    for (unsigned EltNo : indices(NewAllocations))
-      B.createStore(SI->getLoc(),
-                    createAggProjection(B, SI->getLoc(), SI->getSrc(), EltNo),
-                    NewAllocations[EltNo],
-                    StoreOwnershipQualifier::Unqualified);
+    SILBuilderWithScope builder(SI);
+    // In ownership mode, create a destructure of the source and use that to
+    // initialize the aggregate.
+    MultipleValueInstruction *destructured = nullptr;
+    if (SI->getFunction()->hasOwnership()) {
+      if (TT)
+        destructured =
+            builder.createDestructureTuple(SI->getLoc(), SI->getSrc());
+      else
+        destructured =
+            builder.createDestructureStruct(SI->getLoc(), SI->getSrc());
+    }
+    for (unsigned eltNo : indices(NewAllocations)) {
+      if (destructured)
+        builder.createStore(SI->getLoc(), destructured->getResult(eltNo),
```

Comment:
This should be an emitStoreValueOperation. I think there is a bunch of conditionality that is not needed if you used emitDestructureValueOperation.